react/ShippingForm.tsx

```diff
@@ -26,87 +19,81 @@ const ShippingForm: React.FC = () => {
   } = OrderForm.useOrderForm()
   const { selectedAddress, deliveryOptions } = OrderShipping.useOrderShipping()
 
-  const selectedDeliveryOptions = deliveryOptions.filter(
-    ({ isSelected }) => isSelected
-  )
-
-  const [currentStage, setCurrentStage] = useState<ShippingStage>(() => {
-    if (selectedAddress != null) {
-      return ShippingStage.SELECT_DELIVERY_OPTION
-    }
-    if (availableAddresses?.length !== 0) {
-      return ShippingStage.SELECT_ADDRESS
-    }
-    return ShippingStage.CREATE_ADDRESS
+  const { matches, send, state } = useShippingStateMachine({
+    availableAddresses: availableAddresses ?? [],
+    selectedAddress,
+    deliveryOptions,
   })
 
-  const handleAddressCreated = useCallback(() => {
-    setCurrentStage(ShippingStage.SELECT_DELIVERY_OPTION)
-  }, [])
+  const handleAddressCreated = useCallback(
+    (address: Address) => {
+      send({ type: 'SUBMIT_CREATE_ADDRESS', address })
+    },
+    [send]
+  )
 
-  const handleDeliveryOptionSelect = () => {
-    if (selectedAddress.addressType == null) {
-      setCurrentStage(ShippingStage.COMPLETE_ADDRESS)
-    }
+  const handleDeliveryOptionSelect = (deliveryOptionId: string) => {
+    send({ type: 'SUBMIT_SELECT_DELIVERY_OPTION', deliveryOptionId })
   }
 
-  const handleAddressCompleted = ({
-    buyerIsReceiver,
-  }: {
+  const handleAddressCompleted = (
+    updatedAddress: Address,
     buyerIsReceiver: boolean
-  }) => {
-    if (!buyerIsReceiver) {
-      setCurrentStage(ShippingStage.EDIT_RECEIVER_INFO)
-    } else {
-      // go to next step
-    }
+  ) => {
+    send({ type: 'SUBMIT_COMPLETE_ADDRESS', buyerIsReceiver, updatedAddress })
   }
 
-  switch (currentStage) {
-    case ShippingStage.COMPLETE_ADDRESS: {
+  switch (true) {
+    case matches('completeAddress'): {
       return (
         <AddressCompletionForm
           address={selectedAddress}
-          onShippingOptionEdit={() =>
-            setCurrentStage(ShippingStage.SELECT_DELIVERY_OPTION)
-          }
+          deliveryOptions={state.context.deliveryOptions}
+          onShippingOptionEdit={() => send('GO_TO_SELECT_DELIVERY_OPTION')}
           onAddressCompleted={handleAddressCompleted}
+          isSubmitting={matches({ completeAddress: 'submitting' })}
         />
       )
     }
-    case ShippingStage.EDIT_RECEIVER_INFO: {
+    case matches('editReceiverInfo'): {
       return (
-        <ReceiverInfoForm
-          onReceiverInfoSave={() =>
-            setCurrentStage(ShippingStage.SELECT_DELIVERY_OPTION)
-          }
-        />
+        state.context.selectedAddress && (
```

Comment:
is it possible for the user to be editing the receiver info without a selected address?

Reply:
It shouldn't be possible to